File: Data Structures & Algorithms/count-number-of-islands/submission-1.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        int res = 0;
        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                if(grid[i][j] == '1') {
                    res++;
                    dfs(grid, i, j, m, n);
                }
            }
        }
        return res;
    }

private:
    void dfs(vector<vector<char>>& grid, int i, int j, int m, int n) {
        if (
            i < 0 or i >= m
            or j < 0 or j >= n
            or grid[i][j] != '1'
        )
            return;
        
        grid[i][j] = 'X';
        int delta[][2] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        for(auto& d: delta) {
            dfs(grid, i + d[0], j + d[1], m, n);
        }
        return;
    }
};

File: Data Structures & Algorithms/burst-balloons/submission-1.py
class Solution:
    def maxCoins(self, nums: List[int]) -> int:
        nums = [1] + nums + [1]
        dp = {}

        def dfs(l, r):
            if l > r:
                return 0
            if (l, r) in dp:
                return dp[(l, r)]
            
            res = 0
            for i in range(l, r + 1):
                temp = nums[l - 1] * nums[i] * nums[r + 1]
                temp += dfs(l, i - 1)
                temp += dfs(i + 1, r)
                res = max(res, temp)
            
            dp[(l, r)] = res
            return dp[(l, r)]
        
        return dfs(1, len(nums) - 2)
            